### PR TITLE
Put spline SPO multi walker memory to shared resource and split QMCHamiltonian flex_ functions

### PR DESCRIPTION
**Pull request by ye-luo**
Review after #2952
## Proposed changes
1. Put spline SPO multi walker memory to shared resource
2. Put DiracDeterminantBatched multi walker memory to shared resource
3. Split QMCHamiltonian flex_ functions like #2952

1, 2 makes memory consumption mostly under control. Users should not see large memory (MB) increase as DMC progress. Still a few small buffers (KB) needs to be handled but later.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/QMCWaveFunctions/BsplineFactory/OffloadMultiWalkerMem.h
Comment (PDoakORNL) [inline src/QMCWaveFunctions/BsplineFactory/OffloadMultiWalkerMem.h:28]: If I am a scope takes an OffloadMultiWalkerMem as an argument
what can I assume about state of the OMWM.

I'd assume mw_ratios_private, rg_private, mw_pos_copy might contain persistent data written to by some scopes maybe read by others. 

I'd assume mw_offload_scratch, mw_results_scratch, buffer_H2D I can just use for whatever I want to write transfer H2D read/write some more and then transfer D2H read.

det_ratios_buffer_H2D?

File: src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp
Comment (PDoakORNL) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp:392]: I'd rather save the vertical space than horizontal but that's your call.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp
Comment (PDoakORNL) [inline src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp:808]: maybe an #ifndef or NDEBUG assert to avoid the conditional?

File: src/QMCWaveFunctions/Fermion/MatrixDelayedUpdateCUDA.h
Comment (PDoakORNL) [inline src/QMCWaveFunctions/Fermion/MatrixDelayedUpdateCUDA.h:379]: Ok I see this is a testing convenience behavior.  Rather see this in a test helper of some sort. Figuring out how to do that easily is a separate task.

**Review by PDoakORNL — APPROVED** [review]
I think this can go in.

My Concerns about OffloadMultiWalkerMem and checking and allocating for testing resources in the class is something we should add to a punch list for a second pass design review after the dust has settled on shared resource refactoring. Where should we start collecting that list?

**Comment by ye-luo** [comment]
Will deal with the review comments after merging to unblock beta testers.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp:392]: In this case, I prefer bring phi_leader.mw_mem_->det_ratios_buffer_H2D down to det_ratios_buffer_H2D.

File: src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp
Comment (ye-luo) [reply src/QMCWaveFunctions/BsplineFactory/SplineC2ROMPTarget.cpp:808]: This is not a stopper but should allow continue running.

**Pull request merged**
